### PR TITLE
CSHARP-5305: Improve error message when custom LINQ extension method is executed client side.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/MongoEnumerable.cs [hunk 1]
```diff
@@ -42,7 +42,7 @@ internal static HashSet<T> ToHashSet<T>(this IEnumerable<T> source)
         /// <returns>Only meant to be used in Update specifications.</returns>
         public static TSource AllElements<TSource>(this IEnumerable<TSource> source)
         {
-            throw new NotSupportedException("This method is not functional. It is only usable in conjunction with MongoDB.");
+            throw new NotSupportedException($"{nameof(AllElements)} can only be used in LINQ queries executed server-side.");
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/MongoEnumerable.cs:45]: Is this error message good enough or can you think of even better?

File: src/MongoDB.Driver/Linq/MongoEnumerable.cs [hunk 2]
```diff
@@ -862,7 +862,7 @@ public static IEnumerable<TResult> TopN<TSource, TKey, TResult>(
         /// <returns>The filtered results.</returns>
         public static IEnumerable<TSource> Where<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate, int limit)
         {
-            throw new NotSupportedException("This method is not functional. It is only usable in conjunction with MongoDB.");
+            throw new NotSupportedException($"{nameof(Where)} can only be used in LINQ queries executed server-side.");
```
Comment (sanych-sun) [inline src/MongoDB.Driver/Linq/MongoEnumerable.cs:865]: Can move relace by single helper method to throw:
```
private void ThrowNotSupportedException([CallerMemberName]string callerName = null)
{
    throw new NotSupportedException($"{callerName} can only be used in LINQ queries executed server-side.");
}
```
So each method could be converted to simple expression body method:
```
public static IEnumerable<TSource> Where<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate, int limit) 
=> ThrowNotSupportedException()
```

File: src/MongoDB.Driver/Linq/MongoEnumerable.cs (hunk 1 above)
Comment (BorisDog) [reply src/MongoDB.Driver/Linq/MongoEnumerable.cs:45]: IMHO it's good enough.

File: src/MongoDB.Driver/Linq/MongoEnumerable.cs (hunk 2 above)
Comment (BorisDog) [reply src/MongoDB.Driver/Linq/MongoEnumerable.cs:865]: Good idea.

Comment (damieng) [reply src/MongoDB.Driver/Linq/MongoEnumerable.cs:865]: I think losing "MongoDB" is a step backward as no other LINQ providers will be able to translate this.

Comment (rstam) [reply src/MongoDB.Driver/Linq/MongoEnumerable.cs:865]: Good point. I will include "MongoDB" in the error message.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Misc/CustomLinqExtensionMethodHelper.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+using System;
+using System.Runtime.CompilerServices;
+
+namespace MongoDB.Driver.Linq.Linq3Implementation.Misc
+{
+    internal static class CustomLinqExtensionMethodHelper
+    {
+        public static Exception CreateNotSupportedException([CallerMemberName] string methodName = null)
+        {
+            throw new NotSupportedException($"{methodName} can only be used in MongoDB LINQ queries executed server-side.");
```
Comment (sanych-sun) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Misc/CustomLinqExtensionMethodHelper.cs:25]: Should we return it instead of throwing?

**Review by sanych-sun — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver/Linq/DateTimeExtensions.cs [hunk 4]
```diff
@@ -32,7 +33,7 @@ public static class DateTimeExtensions
         /// <returns>The resulting DateTime.</returns>
         public static DateTime Add(this DateTime @this, long value, DateTimeUnit unit)
         {
-            throw new InvalidOperationException("This DateTime.Add method is only intended to be used in LINQ queries.");
+            throw CustomLinqExtensionMethodHelper.CreateNotSupportedException();
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/DateTimeExtensions.cs:36]: Kind of "Yes", but being consistent here probably better. I think we should accept this change.

**Review by sanych-sun — APPROVED** [review]
LGTM

**Pull request merged**